### DISCUSSION TITLE
[Feature request] 是否可以允许 在 Controller 和 Action上配置忽略中间件或守卫？

### DISCUSSION THREAD

**Discussion by flyingcrp** [midwayjs/midway#3372] [Ideas]
### The feature, motivation（功能、动机）

在应用开发中使用中间件和守卫时，难免会遇到需要不受中间件和守卫约束的路由;
在当前 midwayjs 中[中间件的使用方法](https://www.midwayjs.org/docs/middleware#%E4%BD%BF%E7%94%A8%E4%B8%AD%E9%97%B4%E4%BB%B6)和[使用守卫](https://www.midwayjs.org/docs/guard#%E4%BD%BF%E7%94%A8%E5%AE%88%E5%8D%AB)都存在一个问题是：

**无法对已生效的守卫或中间件进行手动忽略**

比如我在 Controller 中使用了中间件或守卫，我无法再该 Controller 下的 Action 上忽略该中间件或守卫；
我无法在特定的Controller 和 Action 上忽略全局中间件和守卫；

要面对这种情况，通常有以下三种常规方法：
1. 全局中间件和守卫中通过 `ignore` 实现，把某个 Controller 下的所有路由或某个固定路由进行排除(还有无法处理的动态路由)
2. 手动引用，在当前的 Controller 中的每一个 Action 手动写，在不需要使用的 Action 上不写；放弃全局中间件改为全手动
3. 使用全局中间件，但抽取一个`public`的辅助路由 Controller 把所有不需要使用全局中间件的 Action 都写在这个公共的控制器里面；
4. 
最后我们又面临如下问题：

第一种方式的问题是，`ignore`会随着这种特殊的`action`的增长变得复杂，同时需要有额外的维护负担；
第二种方法的问题在于，需要更多的重复的引用内容，比如整个 Controller 有十个 Action 但是只有1-2 个Action 是不需要执行中间件的，为了这少部分的 Action 我不得不在其他大多数 Action 上手动使用中间件，如果恰好你还有全局的中间件将更加复杂；
第三种也是当下最常见的一种方法，会通过统一 Controller 或 前缀的方式让中间件绕过该路由，从而可以享受全面中间件带来的便捷；但会产生新的问题：被作为这个控制器下的路由将不再有明确的模块归属，也没有业务归属。这个控制器将接收所有为了绕过全局中间件的所有的 Action； 但这些路由本应该在它所属的模块/功能之下(包括路由前缀)；

因此我们是否可以让 Controller 与 Action 所执行的中间件和守卫能独立分开？以此来解决当前无法分开产生的上诉问题？比如：**允许在Controller 和 Action 上忽略中间件**；
以此来解决  相同业务模块中不同的 Action 或 Controller 不再需要拆分的同时，还能满足其访问控制。

**Comment by czy88840616** [comment]
这种？

```ts

class AMiddleware {

}

@Controller('/', {
   middleaware: [ignorePattern(AMiddleware, '/some_router')]
})
class HomeController {

  @Get('/some_router')
  async route() {}
}

```

**Reply by flyingcrp** [reply]
```
class AMiddleware {

}

@Controller('/', {
   ignoreMiddleaware: [AMiddleware ]
})
class HomeController {
  @Get('/some_router',{  ignoreMiddleaware: [AMiddleware ]})
  async route() {}
}
```

可以这样会更轻便简洁?

**Reply by czy88840616** [reply]
是这样吧。。。很绕的

```ts
@Controller('/', {
   middleaware: [AMiddleware ]
})
class HomeController {
  @Get('/some_router',{  ignoreMiddleaware: [AMiddleware ]})
  async route() {}
}
```

这个东西很难实现的原因在于，中间件逻辑是把类变为方法执行的，执行的时候已经不存在类本身，但是上述的逻辑很明显又是一个类匹配的关系。

**Reply by flyingcrp** [reply]
中间件的缩小范围由 全局 -> controller -> action

这种方式更多的还是解决全局中间件和守卫下,如何给特定的route和action开口子的问题,同时还需要能保证他们依然在他们原属的模块和路由下.

所以实际上在controller和action中都可以会更好.
如果使用了全局,我只需要放开某个action,我直接在具体的action上ignore 就行.

例子1, 使用了全局中间件, controller层级不使用该中间件,则controller及以下均不使用, 控制器级别忽略和在中间件ignore中匹配前缀逻辑一致
```
// 控制器级别忽略
@Controller('/',{  ignoreMiddleaware: [GlobalMiddleware]})
class HomeController {
  @Get('/some_router',)
  async route() {}
}

```
例子2,使用全局中间件, 只需要某些action忽略; 和在中间件ignore中匹配具体路径(user/card/edit)一致. 但是更简单直接
```

// action级别忽略全局,但controller其他依然受全局中间件逻辑控制,这种情况是最多的

@Controller('/')
class HomeController {
 @Get('/some_router1',)
  async route() {}
  @Get('/some_router',{  ignoreMiddleaware: [GlobalMiddleware ]})
  async route() {}
}
```

例子3 不使用全局中间件,由控制器单独使用,但控制器下action 不使用

```
@Controller('/',{  ignoreMiddleaware: [GlobalMiddleware ],useMiddleware:[BMiddleware]})
class HomeController {
 @Get('/some_router1',)
  async route() {}
  @Get('/some_router',{  ignoreMiddleaware: [BMiddleware]})
  async route() {}
}
```

**Reply by czy88840616** [reply]
这个事情和写法无关，和实现以及性能有关，看起来和之前实现guard的时候，要求在中间件里就提前判断走不走路由，是一个事情。

从性能角度来说，如果在路由加一些装饰器（Valid什么的），看起来很简单，实际上底层逻辑就会变成在中间件执行的时候，做全路径扫描正则匹配，找到在忽略列表中的路由，则跳过执行，这个循环再匹配的过程，性能是非常差的。

所以从逻辑上来说，中间件无法感知是否会进到路由中（路由 koa-router 本身就是一个中间件），逻辑不执行到路由中间件是无法知道是否进路由的，所以这是悖论。

**Reply by flyingcrp** [reply]
通过改造已经能满足我期望的效果了.  感谢
